### ISSUE TITLE
Cosine annealing learning rate schedule

### ISSUE THREAD

**Issue by maxmarketit** [facebookresearch/detectron2#1224]
## 🚀 Feature
Cosine learning rate decay as desribed in https://www.dlology.com/blog/bag-of-tricks-for-image-classification-with-convolutional-neural-networks-in-keras/

## Motivation & Examples
We can expect massive training speed-up.
It would be something like setting a configuration like
the highest learning, the lowest learning, and 
how many iterations for one cycle of cosine.

![image](https://user-images.githubusercontent.com/58681175/79367667-35ea0e00-7f89-11ea-911d-0bb0b6561fe2.png)

## Note

We only consider adding new features if they are relevant to many users.

If you request implementation of research papers --
we only consider papers that have enough significance and prevalance.

We do not take requests for most projects in the `projects/` directory,
because they are research code release that is mainly for other researchers to reproduce results.

Instead of adding features inside detectron2,
you can implement many features by [extending detectron2](https://detectron2.readthedocs.io/tutorials/extend.html).
The [projects/](https://github.com/facebookresearch/detectron2/tree/master/projects/) directory
contains many of such examples.

**Comment by ppwwyyxx** [comment]
You can already use `LR_SCHEDULER=WarmupCosineLR`.

You can also use any other LR scheduler you implemented with detectron2 models.

**Comment by maxmarketit** [comment]
Correct me if i am wrong, but WarmupCosineLR does not work for more than 1 cycle of cosine.

  check the code,

        return [
            base_lr
            * warmup_factor
            * 0.5
            * (1.0 + math.cos(math.pi * self.last_epoch / self.max_iters))
            for base_lr in self.base_lrs
        ]

**Comment by ppwwyyxx** [comment]
That's right. You can use a custom scheduler then.

**Comment by rbgirshick** [comment]
> We can expect massive training speed-up

I think this is unlikely, but if you end up trying it and find a massive training speedup please let us know.

**Comment by maxmarketit** [comment]
to ppwwyyxx/

You mean `torch.optim.lr_scheduler`?

If so, I might be missing something here.

Because I only know the training going like the below.

cfg.LR_SCHEDULER_NAME = 'WarmupMultiStepLR' 
cfg.WARMUP_FACTOR = 1.0/1000 
cfg.SOLVER.BASE_LR = 0.02
cfg.SOLVER.MAX_ITER = 300    
trainer = DefaultTrainer(cfg)

I do not see any way to use my custom lr_scheduler unless it is incorporated into 
`detectron2/detectron2/solver/lr_scheduler.py`.
If I can, can you point to the tutorial?

**Comment by ppwwyyxx** [comment]
see https://detectron2.readthedocs.io/tutorials/training.html
you can return your own scheduler in a subclass's https://detectron2.readthedocs.io/modules/engine.html#detectron2.engine.defaults.DefaultTrainer.build_lr_scheduler

**Comment by oakaloverjet** [comment]
@maxmarketit  have you figured out how to implement cosine annealing with detectron2 if so could you please share.
